runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java

```diff
@@ -358,19 +376,19 @@ public void insert(@ConflictParameter K key, V value) {
      */
     @Accessor
     public List<V> scanAndFilter(Predicate<? super V> valuePredicate) {
-        return mainMap.entryStream()
+        return pool.submit(() -> mainMap.entryStream()
                 .map(Entry::getValue).filter(valuePredicate)
-                .collect(Collectors.toCollection(ArrayList::new));
+                .collect(Collectors.toCollection(ArrayList::new))).join();
     }
 
     /** {@inheritDoc} */
     @Override
     @Accessor
     public Collection<Map.Entry<K, V>> scanAndFilterByEntry(
             Predicate<? super Map.Entry<K, V>> entryPredicate) {
-        return mainMap.entryStream()
+        return pool.submit(() -> mainMap.entryStream()
```

Comment:
I know i have seen this problem, but is there a way to create a test that will cause this deadlock. I am not holding the review for this but just want to see if something is possible.

Reply:
Its a timing issue and there isn't a lot of control over how threads access the VLO. 

Let me think about how this can be tested.

Reply:
Without a test case it is hard to be convinced. 
A small test may help such as with a thread pool of 4 and each start with try to hold a common lock, then the successful one will try to do parallel work, this could lead to a possible deadlock.

Reply:
This race condition is really tricky to reproduce because of timing. I wasn't able to do it with latches. 

Apparent, the usage of parallel stream with the common fork-join pool has unspecified behavior
https://stackoverflow.com/a/29272776